azure-storage-queue/src/main/scala/akka/stream/alpakka/azure/storagequeue/AzureQueueSourceStage.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.azure.storagequeue
+
+import com.microsoft.azure.storage.queue.{CloudQueue, CloudQueueMessage}
+import akka.stream.stage.{GraphStage, GraphStageLogic, OutHandler}
+import akka.stream.{Attributes, Outlet, SourceShape}
+import scala.collection.mutable.Queue
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+case class AzureQueueSourceSettings(
+    waitTimeSeconds: Int,
+    maxBufferSize: Int,
+    maxBatchSize: Int
+) {
+  require(maxBatchSize <= maxBufferSize, "maxBatchSize must be lower or equal than maxBufferSize")
+}
+
+object AzureQueueSourceSettings {
+  val default = AzureQueueSourceSettings(20, 100, 10)
+}
+
+class AzureQueueSourceStage(cloudQueue: CloudQueue, settings: AzureQueueSourceSettings)
+    extends GraphStage[SourceShape[CloudQueueMessage]] {
+  val out: Outlet[CloudQueueMessage] = Outlet("AzureCloudQueue.out")
+  override val shape: SourceShape[CloudQueueMessage] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    val buffer = new Queue[CloudQueueMessage]
+
+    private val maxConcurrency = settings.maxBufferSize / settings.maxBatchSize
+    private var maxCurrentConcurrency = maxConcurrency
+    private var currentRequests = 0
+
+    private val successCallback = getAsyncCallback(handleSuccess)
+    private val failureCallback = getAsyncCallback(handleFailure)
+
+    private def canReceiveNewMessages: Boolean = {
+      val currentFreeRequests = (settings.maxBufferSize - buffer.size) / settings.maxBatchSize
+      currentFreeRequests > currentRequests && maxCurrentConcurrency > currentRequests
+    }
+
+    def receiveMessages(): Unit =
+      if (canReceiveNewMessages) {
+        implicit val executionContext = materializer.executionContext
+        currentRequests += 1
+
+        Future {
+          import scala.collection.JavaConverters._
+          val res = cloudQueue.retrieveMessages(settings.maxBatchSize, settings.waitTimeSeconds, null, null)
```

Comment:
Maybe I miss a detail, but I think the process is like this:

(Assume for the sake of argument that the azure queue contains enough messages.)

- The Source is created.
- The Source is pulled from for the first time, i.e., `onPull`-handler is called.
- `onPull` calls `receiveMessage`, and this functions forks off a `Future` waiting for messages from Azure.
- If the future is successful it will receive `maxBatchSize` (default: 10) messages and schedules a call to `handleSuccess`.
- `handleSuccess` puts the 10 messages in `buffer`, pushes 1 message out, and forks off a new `Future` (through a call to `receiveMessage`).
  Now there are 9 messages in `buffer`.
- Another Message is being pulled, i.e., `onPull`-handler is called:
  Since `buffer` is not empty a message is pushed, ***and*** `receiveMessage` is called. 
This call forks off a second `Future`.
  Now there are 8 message in `buffer`.
- This continues until the buffer is empty.
  At this time 10 many futures are running.
  (Assuming that they do not have completed yet.)